File: robot_sdk/c/examples/test_robot_udp.c
/**
 * Example test robot using the Robomesh C SDK over UDP.
 *
 * Demonstrates: authenticate over UDP → heartbeat → messaging.
 * Uses the pre-seeded example-001 robot.
 *
 * Usage:
 *   ./test_robot_udp [host] [udp_port]
 *   Default: localhost:5001
 */

#include "robomesh_udp.h"
#include "robomesh.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

static volatile int running = 1;

static void sigint_handler(int sig) {
    (void)sig;
    running = 0;
}

int main(int argc, char *argv[]) {
    const char *host = argc > 1 ? argv[1] : "localhost";
    int udp_port = argc > 2 ? atoi(argv[2]) : 5001;

    /* Load the pre-seeded test robot's key */
    robomesh_keypair_t kp;
    robomesh_load_keypair("c55608b70c4a9f3b43bd1d23e86aaf4c3b2f4b823f54dc34ac668e85363ef2e1",
                           &kp);

    char pub_hex[65];
    robomesh_public_key_hex(&kp, pub_hex);
    printf("Robot: example-001\n");
    printf("Public key: %s\n", pub_hex);
    printf("Protocol: UDP on %s:%d\n\n", host, udp_port);

    /* Create UDP client */
    robomesh_udp_config_t config = {
        .uuid = "example-001",
        .host = host,
        .udp_port = udp_port,
        .keypair = &kp,
    };
    robomesh_udp_client_t *client = robomesh_udp_create(&config);
    if (!client) {
        fprintf(stderr, "Failed to create UDP client\n");
        return 1;
    }

    /* Authenticate over UDP */
    printf("Authenticating over UDP...\n");
    robomesh_err_t err = robomesh_udp_authenticate(client);
    if (err != ROBOMESH_OK) {
        fprintf(stderr, "UDP auth failed: %s\n", robomesh_udp_last_error(client));
        robomesh_udp_destroy(client);
        return 1;
    }
    const char *jwt = robomesh_udp_get_jwt(client);
    printf("Authenticated! JWT: %.20s...%s\n", jwt, jwt + strlen(jwt) - 10);

    /* Send a test message */
    printf("Sending test message...\n");
    err = robomesh_udp_send(client, "hello from UDP test robot");
    if (err != ROBOMESH_OK) {
        fprintf(stderr, "Send failed: %s\n", robomesh_udp_last_error(client));
    } else {
        printf("Message sent OK\n");
    }

    /* Send heartbeats */
    signal(SIGINT, sigint_handler);
    printf("Sending heartbeats over UDP. Press Ctrl+C to stop.\n");

    while (running) {
        err = robomesh_udp_send_heartbeat(client, NULL, 60);
        if (err != ROBOMESH_OK) {
            fprintf(stderr, "Heartbeat error: %s\n", robomesh_udp_last_error(client));
            break;
        }
        printf("UDP heartbeat OK\n");
        sleep(25);
    }

    printf("Shutting down...\n");
    robomesh_udp_destroy(client);
    return 0;
}

File: robot_sdk/python/examples/test_robot_udp.py
#!/usr/bin/env python3
"""Example test robot using the Robomesh SDK over UDP.

Demonstrates the full UDP lifecycle:
1. Generate keys (or use existing ones)
2. Provision via admin API
3. Authenticate via UDP challenge-response
4. Send heartbeats over UDP
5. Exchange messages over UDP

Usage:
    python test_robot_udp.py

    # Custom server:
    ROBOMESH_HOST=192.168.1.50 ROBOMESH_UDP_PORT=5001 python test_robot_udp.py
"""

import logging
import os
import signal
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from robomesh_sdk import RobotUDPClient, generate_ed25519_keypair
from robomesh_sdk.admin import AdminClient

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(name)s] %(message)s")
logger = logging.getLogger("test_robot_udp")

HOST = os.environ.get("ROBOMESH_HOST", "localhost")
UDP_PORT = int(os.environ.get("ROBOMESH_UDP_PORT", "5001"))
HTTP_PORT = int(os.environ.get("ROBOMESH_HTTP_PORT", "8080"))


def main():
    # 1. Generate a fresh keypair
    _, priv_hex, pub_hex = generate_ed25519_keypair()
    robot_uuid = "test-robot-udp-demo"
    device_type = "test_robot"

    logger.info("Robot UUID:    %s", robot_uuid)
    logger.info("Public key:    %s", pub_hex)

    # 2. Provision via admin API
    admin = AdminClient(host=HOST, http_port=HTTP_PORT)
    admin.login("admin", "password1")
    try:
        admin.provision_robot(robot_uuid, pub_hex, device_type)
        logger.info("Robot provisioned successfully")
    except Exception as e:
        logger.warning("Provision failed (may already exist): %s", e)

    # 3. Connect and authenticate over UDP
    client = RobotUDPClient(
        uuid=robot_uuid,
        private_key_hex=priv_hex,
        host=HOST,
        udp_port=UDP_PORT,
    )

    jwt = client.authenticate()
    logger.info("Authenticated via UDP! JWT: %s...%s", jwt[:20], jwt[-10:])

    # 4. Send heartbeats in background
    client.start_heartbeat(interval=25, ttl=60)

    # 5. Send some messages
    client.send("hello from UDP test robot")
    time.sleep(1)
    client.send('{"type": "status", "battery": 95, "protocol": "udp"}')

    # 6. Listen for incoming packets
    def on_packet(pkt):
        logger.info("Received from server: %s", pkt)

    client.on_message(on_packet)

    # 7. Wait for Ctrl+C
    logger.info("UDP robot running. Press Ctrl+C to stop.")

    def handle_sigint(sig, frame):
        logger.info("Shutting down...")
        client.stop_heartbeat()
        client.disconnect()
        sys.exit(0)

    signal.signal(signal.SIGINT, handle_sigint)

    while True:
        time.sleep(1)


if __name__ == "__main__":
    main()
